Feature/remove diff api

linkdotnet opened the pull request:
This PR removes the Diff API.

Open Points:
- [ ] How do we want to track breaking changes? Is a `BREAKING_CHANGES.md` or `Migration.md` enough? The former just describes what we did - the latter one instructs / advises what to do in `v2`
- [ ] Do we want to delete it right away in the docs? (Personal preference: we do an doc overhaul at the end based on the documents described in point 1).

Your thoughts?

egil commented:
> This PR removes the Diff API.
> 
> Open Points:
> 
> * [ ]  How do we want to track breaking changes? Is a `BREAKING_CHANGES.md` or `Migration.md` enough? The former just describes what we did - the latter one instructs / advises what to do in `v2`

I think we can make do with a Migration.md on the website that has a section for each breaking change, and how to migrate (if possible). The v2 release notes can then link to the migration page on the website.

> * [ ]  Do we want to delete it right away in the docs? (Personal preference: we do an doc overhaul at the end based on the documents described in point 1).

I am a big fan of updating the docs as part of each PR that is merged in. Otherwise, we end up with a massive "docs update task" later.

Wrt. the content of this PR, I am sure there are some users that actually like this feature, so while I do not want to keep the diff and snapshot stuff around in the built-in `RenderedFragment` type, I do think we need to support a way for users to extend that type with their own overrides that enables them to do things like this, i.e. "capture a snapshot of the markup after each render".

That can come as part of a later PR/issue, since that will need design first. But lets create the issue and link it here.

egil reviewed (dismissed):
LGTM. I have not downloaded and verified locally, so I guess there may be other related bits that can be deleted.

linkdotnet commented:
I added the migration guide and deleted the parts in the documentation. 

I fiddled around with how we can keep two versions alive on the same page - but that seems rather complicated with docfx. Sooner or later we have to come up with a solution.

In the docs we can also prominently present the migration guide so users know what to do. Still I think and hope the changes are manageable and only this PR has no direct one-to-one replacement.

linkdotnet commented:
Now we are going a bit off-topic, but what we might pull off to have the documentation for both versions is having a new repository just for the docs itself and having the latest `v1` and `v2` as submodules linked. 
1. Create new repository for the docs
2. Have doc folder for each major version (docs_10, docs_20)
3. Each doc folder has their own docfx.json
4. Add the git submodules:
```bash
git submodule add -b release/v1 bUnitUrl docs_v10/src
git submodule add -b release/v2 bUnitUrl docs_v20/src
```
5. We can generate the docs like this:
```bash
docfx docfx_10.json
docfx docfx_20.json
```
6. Add version toggle - Here we might have to change somehow the template if this is possible (https://dotnet.github.io/docfx/docs/template.html?tabs=modern)

linkdotnet commented:
> I do think we need to support a way for users to extend that type with their own overrides that enables them to do things like this...

They can do this now with the `OnMarkupUpdated` event of the `IRenderFragment`. Sure it will be less convenient but we could even show an example of how to do that in the migration guide.

egil commented:
> > I do think we need to support a way for users to extend that type with their own overrides that enables them to do things like this...
> 
> They can do this now with the `OnMarkupUpdated` event of the `IRenderFragment`. Sure it will be less convenient but we could even show an example of how to do that in the migration guide.
> 
> PS: I was also thinking of having a poll where users can submit whether or not they use a certain feature.

The problem is that only some parts of the plumbing are there. I think it should be easier to set up a scenario where a user replace our RenderedFragment with their own derived type.

egil commented:
> Now we are going a bit off-topic, but what we might pull off to have the documentation for both versions is having a new repository just for the docs itself and having the latest `v1` and `v2` as submodules linked.
> 
> 1. Create new repository for the docs
> 2. Have doc folder for each major version (docs_10, docs_20)
> 3. Each doc folder has their own docfx.json
> 4. Add the git submodules:
> 
> ```shell
> git submodule add -b release/v1 bUnitUrl docs_v10/src
> git submodule add -b main bUnitUrl docs_v20/src
> ```
> 
> 5. We can generate the docs like this:
> 
> ```shell
> docfx docfx_10.json
> docfx docfx_20.json
> ```
> 
> 6. Add version toggle - Here we might have to change somehow the template if this is possible (https://dotnet.github.io/docfx/docs/template.html?tabs=modern)

I was thinking about something similar. Basically, generate the docs site twice with different base URLs. That should mean that as long as the page names stay the same, a simple drop-down can toggle between versions by simply replacing the v1/v2 in the url.

egil commented on MIGRATION.md:
```diff
@@ -0,0 +1,7 @@
+# Migration Guide `v1` to `v2`
+This document describes the changes that need to be made to migrate from bUnit 1.x to 2.x.
+
+## Removal of `GetChangesSinceFirstRender` and `GetChangesSinceLastRender` methods
+The `GetChangesSinceFirstRender` and `GetChangesSinceLastRender` methods have been removed from `IRenderedComponent<TComponent>`. There is no one-to-one replacement for these methods, but the general idea is to select the HTML in question via `Find` and assert against that.
+
+Alternatively, the `IRenderFragment` still offers the `OnMarkupUpdated` event, which can be used to assert against the markup after a render.
```
This is good for now. But I guess we should extend this in the future with an alternative if we have one.

linkdotnet commented:
> > > I do think we need to support a way for users to extend that type with their own overrides that enables them to do things like this...
> > 
> > 
> > They can do this now with the `OnMarkupUpdated` event of the `IRenderFragment`. Sure it will be less convenient but we could even show an example of how to do that in the migration guide.
> > PS: I was also thinking of having a poll where users can submit whether or not they use a certain feature.
> 
> The problem is that only some parts of the plumbing are there. I think it should be easier to set up a scenario where a user replace our RenderedFragment with their own derived type.

Yes absolutely - we might even want to showcase that in the docs and reference this in the `Migration.md`.
This also goes hand in hand with the removal of unnecessary abstraction.
